add hooks for botocore, boto3 and gevent

jstewmon opened the pull request:
The docs on contributing new hooks seem to be slightly out of date, so I just tried to mimic how existing hooks (and tests for hooks) are written.

htgoebel commented:
Thanks. Yes, the hook docs did not made it into 3.0. We are working on it. I've put some remarks into the code.

htgoebel commented on PyInstaller/hooks/hook-boto3.py:
```diff
@@ -0,0 +1,8 @@
+from PyInstaller.utils.hooks import collect_data_files, collect_submodules
```
Please add the licence header like in the hooks-files. To all of your hooks.

htgoebel commented on PyInstaller/hooks/hook-boto3.py:
```diff
@@ -0,0 +1,8 @@
+from PyInstaller.utils.hooks import collect_data_files, collect_submodules
+
```
Please add a comment stating what boto3 is, a link to it's homepage and which version you tested.

htgoebel commented on PyInstaller/hooks/hook-botocore.py:
```diff
@@ -0,0 +1,3 @@
+from PyInstaller.utils.hooks import collect_data_files
+
+datas = collect_data_files('botocore')
```
Please add a comment stating what botocore is, a link to it's homepage and which version you tested.

htgoebel commented on PyInstaller/hooks/hook-gevent.py:
```diff
@@ -0,0 +1,4 @@
+from PyInstaller.utils.hooks import collect_submodules
+
+# required for gevent.monkey.patch_all in gevent 1.1
```
Please add a comment stating what gevent is, a link to it's homepage and which version you tested.

htgoebel commented on PyInstaller/hooks/hook-gevent.py:
```diff
@@ -0,0 +1,4 @@
+from PyInstaller.utils.hooks import collect_submodules
+
```
Please add a comment stating what gevent is, a link to it's homepage and which version you tested.

htgoebel commented on PyInstaller/hooks/hook-gevent.py:
```diff
@@ -0,0 +1,4 @@
+from PyInstaller.utils.hooks import collect_submodules
+
+# required for gevent.monkey.patch_all in gevent 1.1
+hiddenimports = (collect_submodules('gevent'))
```
So this hook shouls be `hook-gevent.monkey`, since if - for any reason `gevent.monkey` is not imported, it is not necessary, right?

htgoebel commented on tests/functional/test_libraries.py:
```diff
@@ -489,3 +489,37 @@ def test_pandas_extension(pyi_builder):
         """)
 
 
+def test_botocore(pyi_builder):
```
Please decoratr all tests with `@importorskip(...)`

htgoebel commented on tests/functional/test_libraries.py:
```diff
@@ -489,3 +489,37 @@ def test_pandas_extension(pyi_builder):
         """)
 
 
```
Please put the test alphabetically into the file. THis makes it easer to find them.

htgoebel commented on tests/requirements.txt:
```diff
@@ -42,3 +42,6 @@ matplotlib
 pandas
+botocore
+boto3
+gevent
```
Again, please sort alphabetically.

jstewmon commented:
Thanks for the feedback. I've updated the PR with your suggestions.

I wasn't able to fully comply with your request for the `requirements.txt` and `test_libraries.py` entries to be sorted because the existing entries are not sorted, and I did not want to pollute the PR diff by sorting the existing lists. So, I added the new entries as if the existing lists were sorted.

htgoebel commented:
> So, I added the new entries as if the existing lists were sorted.

That's absolutely okay :-)

htgoebel commented:
The tests fail for Python 3. Seams like  gevent does not yet support Python 3
```
E     File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/gevent/hub.py", line 289
E       except Exception, ex:
E                       ^
E   SyntaxError: invalid syntax
```
You'll need another decorator for this: `@skipif(is_py2)` (mind the import!) or perhaps better: change to `@importorskip(gevent.hub)`.

(And you may want to write a bug-report for gevent :-)

jstewmon commented:
Hmm... gevent 1.1 (still in beta) adds support for python 3. So, if I update requirements.txt to specify `gevent==1.1b6`, the tests pass for python 3.

Can we have separate test requirements files python 3 and python 2.7, or is there another mechanism that allows us to specify a version requirement based on the python version?

nchammas commented:
Just curious: Does this PR potentially address [the issue I'm seeing here](https://github.com/pyinstaller/pyinstaller/issues/1641#issuecomment-151546874)?

htgoebel commented:
@nchammas Maybe. try it out :-)

htgoebel commented:
> Hmm... gevent 1.1 (still in beta) adds support for python 3. So, if I update requirements.txt to specify gevent==1.1b6, the tests pass for python 3.

I don't think this is a good idea, since we will forget to remove the `bxxx` for sure. (If ever, it should be >= 1.1b6).

> Can we have separate test requirements files python 3 and python 2.7, or is there another mechanism that allows us to specify a version requirement based on the python version?

Yes, but it's complicated any ugly. I suggest using `@importorskip(gevent.hub)` + plus a comment.

jstewmon commented:
It looks like `@importorskip(gevent.hub)` doesn't help in this case because the raised error is a `SyntaxError`, not an `ImportError`.

The hook appears to be unnecessary for `gevent~=1.0.2` - if I remove the `hiddenimports` assignment from the hook, the gevent tests still pass with gevent 1.0.2 on python 2.7.10.

So, I think the best solution at this time is to use a version constraint (`gevent~=1.1b6`) in requirements.txt. This solution also has the benefit that we don't have to remember to remove a `skipif` once version 1.1 is out of beta.

htgoebel commented:
A module having an SyntaxError can not be imported. So `@importorskip` should do its job.

If the hook is not necessary, it should check the version and do nothing for 1.0.

Our aim is not only to pas the tests, but also to do a good job for the users - which includes not including unnecessary stuff if they have a different but commonly used version.

nchammas commented:
@jstewmon - If I want to use PyInstaller on a project that uses boto (not boto3), do I need to submit a patch similar to what you are doing here to make it work?

jstewmon commented:
@nchammas , it doesn't look like boto has any data files, but its utils module has dynamic imports, so it may require a hook. I can write a test case for it, similar to the ones I wrote for botocore and boto3, to see if it appears to require a hook.

nchammas commented:
@jstewmon - I don't want to pile any extra work on you, but if you choose to cover old boto as well as part of this PR that would be fantastic. Otherwise I might take a crack it myself since I'm trying to get PyInstaller to work with my project which is currently still on boto.

jstewmon commented:
@htgoebel , I think your description of how `@importskip` should work is reasonable. Would you like for me to change it from this...

```python
for m in modules:
        try:
            __import__(m)
        except ImportError:
            mods_avail = False
```

...to this...

```python
for m in modules:
        try:
            __import__(m)
        except Exception:
            mods_avail = False
```
...? If so, I can remove the gevent version constraint and instead rely on `@importskip`.

> Our aim is not only to pas the tests, but also to do a good job for the users - which includes not including unnecessary stuff if they have a different but commonly used version.

Fair point. My original solution was intended to provide the most coverage because was was not certain that I knew all the modules that needed to be listed by hiddenimports. I'll update to only include the modules that are dynamically imported by monkey, and only assign those when `is_module_version('gevent', '>=', '1.1b0')` is True.

jstewmon commented:
@nchammas , I added a hook and test for boto. You can use it immediately in your project by adding hook-boto.py to a directory in your project (say, `pyinstaller`), and giving the `--additional-hooks-dir` option to the pyinstaller cli: `pyinstaller --additional-hooks-dir pyinstaller/ ...` (trailing slash seems to be required).

jstewmon commented:
@htgoebel , do you have any feedback on the latest update?

htgoebel commented on tests/functional/scripts/pyi_lib_boto.py:
```diff
@@ -0,0 +1,45 @@
+
+credentials = dict(
+    aws_access_key_id='ASIAIH3F2FU3T63KIXKA',
+    aws_secret_access_key='lnN4qk1a0SuQAFVsGA+Y+ujo2/5rLq2j+a1O4Vuy'
```
Sure you want a secret key here?

jstewmon replied:
Those are fake values. There is some validator that was causing an error like "no auth method available" if I didn't use values of a valid format.

htgoebel commented on PyInstaller/utils/tests.py:
```diff
@@ -41,7 +41,7 @@ def importorskip(*modules):
         try:
             __import__(m)
-        except ImportError:
+        except Exception:
```
I've re-thought this: `Exception` is to wide, as it would catch anything: I suggest using
`except (ImportError, SyntaxError):`

jstewmon replied:
Perfectly reasonable. Fixed in the latest commit.

jstewmon commented:
@htgoebel , thanks for the feedback. If the next build passes, I'll squash the branch, and edit the commit message, so that everything looks good when merged.

nchammas commented:
The AppVeyor build failed. Is that something that needs to be resolved before this can be merged in?

matysek commented:
@nchammas No, nothing to do. We just haven't stabilized tests on appveyor yet.

htgoebel commented:
Thanks again for this pull-request.

jstewmon commented:
You're welcome. I'm happy to have been able to contribute something back to this awesome project!

nchammas commented:
@htgoebel - Just looking at the [release notes for 3.1](https://github.com/pyinstaller/pyinstaller/releases/tag/v3.1)--I think you forgot to mention boto (vs. boto3, which you got) in the list of new hooks.
